js/src/redux/providers/worker.js

```diff
@@ -0,0 +1,72 @@
+// Copyright 2015, 2016 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+import PromiseWorker from 'promise-worker';
+import runtime from 'serviceworker-webpack-plugin/lib/runtime';
+
+import { setWorker } from './workerActions';
+
+function getWorker () {
+  let workerRegistration;
+
+  // Setup the Service Worker
+  if ('serviceWorker' in navigator) {
+    workerRegistration = runtime
+      .register()
+      .then(() => navigator.serviceWorker.ready)
+      .then((registration) => {
+        const worker = registration.active;
+        worker.controller = registration.active;
+
+        return new PromiseWorker(worker);
+      });
+  } else {
+    workerRegistration = Promise.reject('Service Worker is not available in your browser.');
+  }
+
+  return workerRegistration;
+}
+
+export const setupWorker = (store) => {
+  const { dispatch, getState } = store;
+
+  const state = getState();
+  const stateWorker = state.worker.worker;
+
+  if (stateWorker !== undefined && !(stateWorker && stateWorker._worker.state === 'redundant')) {
```

Comment:
This is just because `worker` can be undefined if it hasn't loaded yet, or `null` if not supported